datalad/nonasyncrunner.py

```diff
@@ -0,0 +1,184 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Thread based subprocess execution with stdout and stderr passed to protocol objects
+"""
+
+import logging
+import os
+import queue
+import subprocess
+import threading
+import time
+from typing import Any, Dict, IO, List, Type, Union, Optional
+
+from .cmd import WitlessProtocol
+
+logger = logging.getLogger("datalad.runner")
+
+STDOUT_FILENO = 1
+STDERR_FILENO = 2
+
+
+class _ReaderThread(threading.Thread):
+
+    def __init__(self,
+                 file: IO,
+                 q: queue.Queue,
+                 command: Union[str, List]):
+        """
+        Parameters
+        ----------
+        file:
+          File object from which the thread will read data
+          and write it into the queue. This is usually the
+          read end of a pipe.
+        q:
+          A queue into which the thread writes what it reads
+          from file.
+        command:
+          The command for which the thread was created. This
+          is mainly used to improve debug output messages.
+        """
+        super().__init__(daemon=True)
+        self.file = file
+        self.queue = q
+        self.command = command
+        self.quit = False
+
+    def __str__(self):
+        return f"ReaderThread({self.file}, {self.queue}, {self.command})"
+
+    def request_exit(self):
+        """
+        Request the thread to exit. This is not guaranteed to
+        have any effect, because the thread might be waiting in
+        `os.read()` or on `queue.put()`, if the queue size is finite.
+        To ensure thread termination, you can ensure that another thread
+        empties the queue, and try to trigger a read on `self.file.fileno()`,
+        e.g. by writing into the write-end of a pipe that is connected to
+        `self.file.fileno()`.
+        """
+        self.quit = True
+
+    def run(self):
+        logger.debug("%s started", self)
+
+        while not self.quit:
+
+            data = os.read(self.file.fileno(), 1024)
+            if data == b"":
+                logger.debug("%s exiting (stream end)", self)
+                self.queue.put((self.file.fileno(), None, time.time()))
+                return
+
+            self.queue.put((self.file.fileno(), data, time.time()))
+
+
+def run_command(cmd: Union[str, List],
+                protocol_class: Type[WitlessProtocol],
```

Comment:
uff -- spent some time chasing it in "resolved" comments.  My feeling grow stronger to keep it consistent here with runner itself and have just `protocol` not `protocol_class`.  My comments were:

 I should have provided more context and explicit desire right away: we have it as a protocol for WitlessRunner as well and just document that it should be a class and not an instance. It would be odd to use `protocol=` there and `protocol_class=` here. IMHO it would be better to stay consistent and just keep it as a protocol in the signature (and you could make it into protocol_instance or alike in the body).

Reply:
Very well, `protocol` it is.